### ISSUE TITLE
report hostname and key if generated

### ISSUE THREAD

**Issue by t2d** [systemli/ansible-role-hidden-service#4]
Maybe it would be nice to report hostname and private_key if they get generated.
Thus, one could put them in the host config later on.

**Comment by t2d** [comment]
acc9a475b756b9e386160a9739165ee4a6c42e3b shows only hostname, but that should be enough for most setups

**Comment by xshadow** [comment]
Former implemented solution resulted in a fail of the role on creation or deletion of a host. 

Creation: 

The show hostname handler was executed too fast, for tor key generation after a restart - so no information could be gathered.

Deletion: 
The hostname that should be shown is not present after a deletion. Also no information could be gathered. 

A report of a hostname is a really interesting feature for this role.

**Comment by t2d** [comment]
> The show hostname handler was executed too fast, for tor key generation
> after a restart - so no information could be gathered.
> 
> Then we probably need to chain the handlers

**Comment by conorsch** [comment]
> The show hostname handler was executed too fast, for tor key generation after a restart - so no information could be gathered.

Sounds like that can be handled via a `wait_for` task that targets a filepath.

> The hostname that should be shown is not present after a deletion. Also no information could be gathered. 

The role already checks for "state" on each hidden service, so skipping the display tasks is trivial.

**Comment by xshadow** [comment]
I came up with a solution (https://github.com/systemli/ansible-role-hidden-service/commit/b66cbc1412bb4bdbc676d41362aff42df83380a6) , but the output of Ansible debug task, when iterating over items, is really awful! 

It results in a pretty long output, and you can miss the debug line. Any other ideas? 

```
ok: [hidden-www-xxx.xxx.org] => (item={u'stdout': u'hcwbvqaqti2b45if.onion', u'cmd': [u'cat', u'/var/lib/tor/foobar_website/hostname'], u'end': u'2016-04-29 17:16:20.864476', '_ansible_no_log': False, u'warnings': [], u'changed': True, u'start': u'2016-04-29 17:16:20.858180', u'delta': u'0:00:00.006296', 'item': {'value': {u'hidden_service_ports': [[80, 9997]], u'hidden_service_private_key': None, u'hidden_service_hostname': None}, 'key': u'foobar_website'}, u'rc': 0, 'invocation': {'module_name': u'command', u'module_args': {u'creates': None, u'executable': None, u'chdir': None, u'_raw_params': u'cat /var/lib/tor/foobar_website/hostname', u'removes': None, u'warn': True, u'_uses_shell': False}}, 'stdout_lines': [u'hcwbvqaqti2b45if.onion'], u'stderr': u''}) => {
    "item": {
        "_ansible_no_log": false, 
        "changed": true, 
        "cmd": [
            "cat", 
            "/var/lib/tor/foobar_website/hostname"
        ], 
        "delta": "0:00:00.006296", 
        "end": "2016-04-29 17:16:20.864476", 
        "invocation": {
            "module_args": {
                "_raw_params": "cat /var/lib/tor/foobar_website/hostname", 
                "_uses_shell": false, 
                "chdir": null, 
                "creates": null, 
                "executable": null, 
                "removes": null, 
                "warn": true
            }, 
            "module_name": "command"
        }, 
        "item": {
            "key": "foobar_website", 
            "value": {
                "hidden_service_hostname": null, 
                "hidden_service_ports": [
                    [
                        80, 
                        9997
                    ]
                ], 
                "hidden_service_private_key": null
            }
        }, 
        "rc": 0, 
        "start": "2016-04-29 17:16:20.858180", 
        "stderr": "", 
        "stdout": "hcwbvqaqti2b45if.onion", 
        "stdout_lines": [
            "hcwbvqaqti2b45if.onion"
        ], 
        "warnings": []
    }, 
    "msg": "A hidden service for /var/lib/tor/foobar_website/ was created or the address is not present in your vars file. It's reachable via hcwbvqaqti2b45if.onion . Don't forget to add hostname and privatekey to your vars file"
}
```

**Comment by conorsch** [comment]
@xshadow I ran into the same problem with overly verbose messages. Apparently it's a common complaint when using debug with a loop: https://github.com/ansible/ansible/issues/5564 Using workarounds to pare back the output, we can get message quite terse:

```
TASK [ansible-role-hidden-service : display hidden service url] ****************
ok: [do-general] => (item=4opabs5s2ot6o7iq.onion) => {
    "item": "4opabs5s2ot6o7iq.onion",
    "msg": "Hidden service URL: 4opabs5s2ot6o7iq.onion"
}
```

There's still a duplicate `item` line when all we want is `msg`, but it's an improvement. The downside is that we need to filter the object lists into the attributes we want to display, which makes listing multiple attributes such as the service name, as in your example, difficult.

The best solution I've found is using a template lookup to format the message more precisely:

```
TASK [ansible-role-hidden-service : display hidden service url] ****************
ok: [do-general] => {
    "msg": "Hidden service URL (ports [22, 22]): 4opabs5s2ot6o7iq.onion\n"
}
```

That trailing newline is quite difficult to remove, but I'm happy leaving it there, because it's better than what the role currently offers. Of course, we can use the same message from your previous example:

```
TASK [ansible-role-hidden-service : display hidden service url] ****************
ok: [do-general] => {
    "msg": "A hidden service for /var/lib/tor/ssh/ was created or the address 
is not present in your vars file. It's reachable via 4opabs5s2ot6o7iq.onion . 
Don't forget to add hostname and privatekey to your vars file.\n"
}
```

Hopefully this is helpful. Happy to push up a feature branch or open a PR if you find one of these approaches acceptable.

**Comment by xshadow** [comment]
Closed via 29eff45ddb71f018039695559895f1f4372ab1a9
